### DISCUSSION TITLE
Regex in patterns

### DISCUSSION THREAD

**Discussion by jpambrun** [ast-grep/ast-grep#378] [Q&A]
This tool looks amazing. I am trying to use cases of the top of my head and one was: find all update() functions called on an object whos name  contains "coll".

something Iike 
``` sh
ast-grep --lang js --pattern '[*coll*].update($$$)'
```
and would match `userColl.update(...)` and `clientCollection.update(...)`

where the part in `[]` would be a regex matcher, maybe it could be capturing too with `$META_NAME`. Is this possible?

**Comment by HerringtonDarkholme** [comment]
Thanks for trying out ast-grep!
The use case is certainly possible. But sadly you cannot easily access them in `pattern` though.  Instead, you can use a rule to describe using both regex and pattern.

<img width="749" alt="image" src="https://github.com/ast-grep/ast-grep/assets/2883231/2e399ef7-c22b-4048-81d5-df13c33155f9">

See
https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6ImphdmFzY3JpcHQiLCJxdWVyeSI6ImNvbnNvbGUubG9nKCRNQVRDSCkiLCJjb25maWciOiJydWxlOlxuICBwYXR0ZXJuOiAkQS51cGRhdGVcbiAgcmVnZXg6IC4qQ29sbCIsInNvdXJjZSI6InVzZXJDb2xsLnVwZGF0ZSh7XG4gIG5hbWU6ICdOZXdOYW1lJ1xufSlcblxuY2xpZW50Q29sbGVjdGlvbi51cGRhdGUoe1xuICB2ZXJzaW9uOiAyXG59KSJ9


To use the rule on command line, you need first save it in a yaml and [add other related](https://ast-grep.github.io/guide/rule-config.html) fields like `id` and `lang`. 

Then you can use the command line to search your code.

```sh
ast-grep scan -r myrule.yml
```

**Reply by HerringtonDarkholme** [reply]
That said, I think using regex to filter meta variable is quite often to use. Having a YAML for it is really heavy. 

Let me think an API to support this use case...

**Comment by andreszapatavff** [comment]
@HerringtonDarkholme Great tool! I'm loving it. 

I have a similar situation, where I want to create a capture group and refernce it in the the fix section.

Basically, I want to transform:

`import {SomeNameController}  from '@screens'` 

to:

`import {SomeNameScreenController}  from '@screens'` 

I have the rule that finds what I need, but I can't figure out how to reference the capture group. 

```
id: screen-controller
language: Tsx
rule:
  kind: import_specifier
  inside:
    kind: named_imports
    regex: (?<NAME>.*)(Controller)
fix: What can I do here??
```

Any suggestions? Thank you!

**Reply by HerringtonDarkholme** [reply]
You can use [transform](https://ast-grep.github.io/guide/rewrite-code.html#use-transform-in-rewrite) here.
More specifically, you can use [`replace`](https://ast-grep.github.io/reference/yaml/transformation.html#replace)

**Reply by HerringtonDarkholme** [reply]
https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6ImphdmFzY3JpcHQiLCJxdWVyeSI6IjwkQSAkJCQgPiQkJDwvJEE+IiwicmV3cml0ZSI6IiIsImNvbmZpZyI6ImlkOiBzY3JlZW4tY29udHJvbGxlclxubGFuZ3VhZ2U6IFRzeFxucnVsZTpcbiAga2luZDogaW1wb3J0X3NwZWNpZmllclxuICBwYXR0ZXJuOiAkTkFNRVxuICBpbnNpZGU6XG4gICAga2luZDogbmFtZWRfaW1wb3J0c1xuICAgIHJlZ2V4OiAoPzxOQU1FPi4qKShDb250cm9sbGVyKVxudHJhbnNmb3JtOlxuICBORVdfQ09OVFJPTExFUjpcbiAgICByZXBsYWNlOlxuICAgICAgYnk6IFNjcmVlbkNvbnRyb2xsZXJcbiAgICAgIHJlcGxhY2U6IENvbnRyb2xsZXJcbiAgICAgIHNvdXJjZTogJE5BTUVcbmZpeDogJE5FV19DT05UUk9MTEVSIiwic291cmNlIjoiaW1wb3J0IHtTb21lTmFtZUNvbnRyb2xsZXJ9ICBmcm9tICdAc2NyZWVucydcblxuIn0=

**Reply by andreszapatavff** [reply]
Thank you!
